### PR TITLE
Adds Slider as a web component

### PR DESCRIPTION
**Pull request by procload**
Adds the Fluent Slider web component

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/71c3165c70d04f3b7e65a9c30eed058a1c7eec6d'>71c3165c70d04f3b7e65a9c30eed058a1c7eec6d</a></sub>
<div id="bundleSizeComment9423"/>

File: packages/web-components/src/slider/slider.spec.md [hunk 1]
```diff
@@ -76,6 +76,7 @@ None
   - `rail` is in Fluent React; `track` is in FAST Foundation
   - No value indicators to the left/right of slider in FAST Slider. These should be present in Fluent version. This will be handled through slotting in the the SliderLabel component.
   - No default support for the ticks on the slider when `step` is set. This will be handled through slotting in the SliderLabel component.
+  - Added attribute of `marks` whose default value is false. This will show/hide the ticks on the slider when `step` is set.
```
Comment (bearcat-msft) [inline packages/web-components/src/slider/slider.spec.md:79]: Where does the language "marks" come from? is this consistent with fluent design language?

File: packages/web-components/src/slider/slider.styles.ts [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    min-height: var(--thumb-size);
+    touch-action: pan-x;
+    padding: 8px 0;
+  }
+
+  :host([disabled]:hover) {
+    cursor: initial;
+  }
+
+  :host(:focus-visible) {
```
Comment (bearcat-msft) [inline packages/web-components/src/slider/slider.styles.ts:63]: Perhaps we should have focus visible styling someplace accessible in the repo so the way in which these styles are applied is consistent throughout all our components?

File: packages/web-components/src/slider/slider.styles.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  .track {
+    align-self: start;
+    position: absolute;
+    background-color: ${colorNeutralStrokeAccessible};
+    border-radius: ${borderRadiusMedium};
+  }
+
+  :host([step]) .track::after {
+    content: '';
+    display: block;
```
Comment (bearcat-msft) [inline packages/web-components/src/slider/slider.styles.ts:139]: I think display block is not required when position is set to absolute.

File: packages/web-components/src/slider/slider.styles.ts [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    position: absolute;
+    border-radius: ${borderRadiusMedium};
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
+    width: 100%;
+    inset: 0 -2px;
```
Comment (bearcat-msft) [inline packages/web-components/src/slider/slider.styles.ts:147]: inset is the shorthand for top, left, right, and bottom. both are not required. https://developer.mozilla.org/en-US/docs/web/css/inset

File: packages/web-components/src/slider/slider.styles.ts (hunk 2 above)
Comment (chrisdholt) [reply packages/web-components/src/slider/slider.styles.ts:63]: My usolicited $0.02 is that we should avoid abstractions unless and until we know it's consistently repeatable. I don't believe that focus styling is or will be applied universally to all elements - is it?

File: packages/web-components/src/slider/slider.styles.ts (hunk 3 above)
Comment (procload) [reply packages/web-components/src/slider/slider.styles.ts:139]: I did not know that!

ogvhs-the-more-you-know

File: packages/web-components/src/progress-bar/progress-bar.styles.ts [hunk 5]
```diff
@@ -13,7 +13,7 @@ import {
   colorPaletteRedBackground3,
 } from '../theme/design-tokens.js';
 
-/** Text styles
+/** ProgressBar styles
```
Comment (chrisdholt) [inline packages/web-components/src/progress-bar/progress-bar.styles.ts:16]: I'm fine if we keep this, but ideally we can handle separate concerns in separate PR's as a policy

File: packages/web-components/src/slider/slider.stories.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+export const Slider = renderComponent(storyTemplate).bind({});
+
+export const SliderVertical = renderComponent(html<SliderStoryArgs>`
+  <fluent-slider orientation="vertical" step="20" value="60" min="0" max="100"></fluent-slider>
+`);
+
+export const SliderSmall = renderComponent(html<SliderStoryArgs>`
+  <fluent-slider size="small" value="10" min="0" max="10"></fluent-slider>
+`);
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.stories.ts:69]: Should this be "size" and include examples of both?

File: packages/web-components/src/slider/slider.stories.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+      },
+    },
+  },
+} as SliderStoryMeta;
+
+export const Slider = renderComponent(storyTemplate).bind({});
+
+export const SliderVertical = renderComponent(html<SliderStoryArgs>`
+  <fluent-slider orientation="vertical" step="20" value="60" min="0" max="100"></fluent-slider>
+`);
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.stories.ts:65]: Should this be "Orientation" and include examples of both?

File: packages/web-components/src/slider/slider.stories.ts [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  <fluent-slider step="10" value="10" min="0" max="100"></fluent-slider>
+`);
+
+export const SliderDisabled = renderComponent(html<SliderStoryArgs>`
+  <fluent-slider disabled value="10" min="0" max="100"></fluent-slider>
+`);
+
+export const SliderDisabledWithSteps = renderComponent(html<SliderStoryArgs>`
+  <fluent-slider step="25" disabled value="50" min="0" max="100"></fluent-slider>
+`);
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.stories.ts:81]: Should these both be under a "Disabled" story?

File: packages/web-components/src/slider/slider.styles.ts [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+ */
+export const styles = css`
+  ${display('inline-grid')} :host {
+    --thumb-size: 18px;
+    --thumb-padding: 3px;
+    --thumb-translate: calc(var(--thumb-size) * -0.5 + var(--track-width) / 2);
+    --track-overhang: -2px;
+    --track-width: 4px;
+    --fast-slider-height: calc(var(--thumb-size) * 10);
+    --slider-direction: 90deg;
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.styles.ts:31]: These end up as global variables - is that desired on the Fluent side of things (does this exist in React as well or is this net-new delta for WC's)?

File: packages/web-components/src/slider/slider.styles.ts [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    transition: all 0.2s ease;
+  }
+
+  :host([orientation='horizontal']) .thumb-container {
+    transform: translateX(calc(var(--thumb-size) * 0.5)) translateY(calc(var(--thumb-translate) * -1.5));
+  }
+
+  :host([orientation='horizontal'][size='small']) .thumb-container {
+    transform: translateX(calc(var(--thumb-size) * 0.5)) translateY(calc(var(--thumb-translate) * -1.35));
+  }
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.styles.ts:86]: Given this is the default i don't think you need to include the added selector for orientation here (do you?)

File: packages/web-components/src/slider/slider.styles.ts [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+  .thumb-cursor:active {
+    background-color: ${colorCompoundBrandBackgroundPressed};
+  }
+  :host([disabled]) .thumb-cursor {
+    background-color: ${colorNeutralForegroundDisabled};
+    box-shadow: inset 0 0 0 var(--thumb-padding) ${colorNeutralBackground1}, 0 0 0 1px ${colorNeutralStrokeDisabled};
+  }
+
+  /* Positioning Region */
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.styles.ts:114]: Just a note, I'm not sure these comments are going to get parsed out.

File: packages/web-components/src/slider/slider.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+ */
+export class Slider extends FASTSlider {
+  /**
+   * The size of the slider
+   * @public
+   * @remarks
+   * HTML Attribute: size
+   */
+  @attr
+  public size: SliderSize = 'medium';
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.ts:17]: Per @miroslavstastny and prior comments, let's keep the defaults set in CSS only, no need to set the default value here.

File: packages/web-components/src/slider/slider.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+  connectedCallback(): void {
+    super.connectedCallback();
+    if (this.step) {
+      const totalSteps = 100 / Math.floor((this.max - this.min) / this.step);
+      this.style.setProperty('--step-rate', totalSteps.toString() + '%');
+    }
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.ts:24]: This is only going to be set once, should this be in the `stepChanged` function? You can confirm the behavior by initializing without steps and then setting the attribute...or changing an existing step attribute. Would we also want to _remove_ this property in the case that step was removed?

File: packages/web-components/src/slider/slider.options.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import { ValuesOf } from '@microsoft/fast-foundation';
+
```
Comment (chrisdholt) [inline packages/web-components/src/slider/slider.options.ts:2]: We should likely re-export the SliderOrientation from Foundation to ensure that everything can be included when bringing this in (no other exports needed).

**Review by chrisdholt — CHANGES_REQUESTED** [review]
Looks good overall, a few things I think we need to clean up/fix. Also, remember to export from the package.json

File: packages/web-components/src/progress-bar/progress-bar.styles.ts (hunk 5 above)
Comment (procload) [reply packages/web-components/src/progress-bar/progress-bar.styles.ts:16]: Yeah, it was super small so I kept it in. But I understand.

File: packages/web-components/src/slider/slider.styles.ts (hunk 9 above)
Comment (procload) [reply packages/web-components/src/slider/slider.styles.ts:31]: Is there a way to scope them to only this component? They're pretty helpful for reducing the amount of CSS required and can come in handy when switching between how attributes affect the appearance.

File: packages/web-components/src/slider/slider.ts (hunk 13 above)
Comment (procload) [reply packages/web-components/src/slider/slider.ts:24]: I might not have enough knowledge here, and I couldn't find an example or text in the docs to answer this question: If I put this in the stepChanged() function how can I set the initial value when the component is initially rendered? Is stepChanged() only called when the attribute is changed post render?

File: packages/web-components/src/slider/slider.options.ts (hunk 14 above)
Comment (procload) [reply packages/web-components/src/slider/slider.options.ts:2]: Would this look like:
```
import { SliderOrientation } from '@microsoft/fast-foundation';

/**
 * Applies orientation to the slider
 * @public
 */
export type SliderOrientation = ValuesOf<typeof SliderOrientation>;
```
Or should I define the options as I did in SliderSize, such as:
```
/**
 * SliderOrientation Constants
 * @public
 */
export const SliderOrientation = {
  horizontal: 'horizontal',
  vertical: 'vertical',
} as const;
```

File: packages/web-components/src/slider/slider.ts (hunk 13 above)
Comment (chrisdholt) [reply packages/web-components/src/slider/slider.ts:23]: @procload my understanding was to not show steps by default - has that changed?

Comment (procload) [reply packages/web-components/src/slider/slider.ts:25]: That's correct. But when I moved the logic into the stepChanged() function, it no longer detects the step attribute until I manually change that value in the DOM via JavaScript (or the Inspector).

Comment (chrisdholt) [reply packages/web-components/src/slider/slider.ts:25]: I'll pull your branch down...to confirm, when you have the below code it's not firing on connect?

```hml
<fluent-slider step="5"></fluent-slider>
```

Comment (chrisdholt) [reply packages/web-components/src/slider/slider.ts:31]: @procload 

```ts
private stepChanged(): void {
    super.stepChanged();

    if (this.step) {
      const totalSteps = 100 / Math.floor((this.max - this.min) / this.step);

      this.style.setProperty('--step-rate', totalSteps.toString() + '%');
    }
  }
```
I see the property being set on the host element. I don't see the styles applying like you might expect, but it's there with the above from what I can see.

Comment (chrisdholt) [reply packages/web-components/src/slider/slider.ts:31]: Probably want to check the math on the above...I'm getting the following in the console for all stories when I log the value of `totalSteps`:
image

File: packages/web-components/src/slider/slider.styles.ts (hunk 9 above)
Comment (procload) [reply packages/web-components/src/slider/slider.styles.ts:31]: @chrisdholt Following up here. Will these be scoped to the component since they are in the :host selector? At least that was my understanding.

**Comment by procload** [comment]
@chrisdholt or @eljefe223 I merged in Chris' Slider code and resolved the previous comments. When you have a chance (I know it might be slow, you're busy), can you take another look?

**Review by chrisdholt — APPROVED** [review]

**Comment by jpaims** [comment]
Design review:

| # | Issue | Category | 
| --- | ----------- | ----------- |  
| 1 | Check focus border style. There should be an additional inner 1px stroke (--colorStrokeFocus1) | Accessibility | 
| 2 | Focus border seems to appear on any mouse interaction but should only appear on keyboard tabbing | Accessibility |  
| 3 | In vertical orientation, steps don't seem to quite match the slider thumb position | Styling | 

Will there be a property for showing min and max values or input fields on the ends of the slider?

**Comment by procload** [comment]
@eljefe223 Not a big rush, but can you take a look when you get a chance? I made some minor design updates per @jpaims after Chris approved.
